src/containers/projects/new/step-2.svelte

```diff
@@ -1,123 +1,177 @@
-  // Somewhat counter-intuitive, but the activityList has nothing to do
-  //   with the concept of internal activities on the backend. Moreover, internal
-  //   activities are not in the activityList, since they are not displayed on the page.
-
-  const dispatch = createEventDispatcher();
-  const unsubscribe = project.subscribe(
-    ({ activities: backendActivities }) => {
-      activityList = synchronizeActivityLists(activityList, backendActivities);
-    },
-  );
-  onDestroy(unsubscribe);
-
   function duplicateActivity({ detail: activity }) {
     const newActivity = copyActivity(activity);
     delete newActivity.id;
-    newActivity._type = ActivityTypes.NEW;
-    activityList.push(newActivity);
+    createActivity(newActivity);
+  }
+
+  async function createActivity(activity) {
+    try {
+      ({
+        id: activity.id,
+        project: activity.project,
+      } = await api.json(api.post(`/projects/${$project.id}/activities`, {
+        data: prepareForBackend(activity),
+        csrfToken: $session.account.csrf_token,
+      })));
+      activity._editing = true;
+      $project.activities.push(activity);
+      project = project;
+    } catch (e) {
+      showSnackbar({
+        props: { text: 'Couldn\'t create the activity, try reloading the page' },
+      });
+      console.error(e);
+    }
     scrollToLast = true;
-    activityList = activityList;
   }
 
-  function deleteActivity({ detail: activity }) {
-    activityList = activityList.filter(
-      act => act.name !== activity.name || (act._type !== ActivityTypes.DISPLAY && act._type !== ActivityTypes.TEMPLATE),
-    );
-    dispatch('delete-activity', activity.id || activity.name);
-    if (activityList.length === 0) {
-      const activity = getBlankActivity();
-      activity._type = ActivityTypes.NEW;
-      activityList.push(activity);
+  async function deleteActivity({ detail: activity }) {
+    try {
+      await api.json(api.del(
+        `/projects/${$project.id}/activities/${activity.id}`,
+        { csrfToken: $session.account.csrf_token },
+      ));
+      $project.activities = $project.activities.filter(act => act.id !== activity.id);
+      if ($project.activities.find(act => !act.internal) == null) {
+        createActivity(getBlankActivity());
+      }
+
+      showSnackbar({
+        props: {
+          text: 'Activity deleted',
+          action: { text: 'undo', callback() { createActivity(activity); } },
+        },
+      });
+    } catch (e) {
+      showSnackbar({
+        props: { text: 'Couldn\'t delete the activity, try reloading the page' },
+      });
+      console.error(e);
+    }
+  }
+
+  function propagateDates({ detail }) {
+    for (let activity of $project.activities) {
+      if (activity.internal) {
+        continue;
+      }
+
+      if (activity.timeframe == null
+          || activity.timeframe.start == null && activity.timeframe.end == null) {
```

Comment:
Use guard clauses. Also, this will not `patch` the date if only one of the `start` or `end` is null.

Reply:
Guard clauses?
As for not patching, that's intentional